### PR TITLE
Add smart reply to remove boilerplate

### LINKED ISSUE #50: Add messenger.send helper to reduce boilerplate
Originally from https://github.com/CondeNast/launch-vehicle-fbm/issues/30#issuecomment-283773917
instead of:

    on('text', ({senderId, text, session}) => messenger.send(senderId, new Text(`echo ${text}`)))

something like:

    on('text', (res) => res.send(new Text(`echo ${res.text}`)))

`res` would be an object containing the same information (`senderId`, `text`, `session`, etc.) but also have a `send` method. This way, it's backwards compatible.

This will eliminate the boilerplate of having to surface `senderId` to send a reply, and post #30, will eliminate needing to dig up the right `pageId` to send as.

### PR DESCRIPTION
**Pull request by crccheck**
## Why are we doing this?

Multitenant bots will have to specify page id when sending a message. To do that now, they would have to do `messenger.send(senderId, responseObj, session._pageId)`, which is very not ideal. Luckily everything has been backwards compatible and the page routing logic has been hidden. This adds another backwards compatible change so bots only have to do `reply(responseObj`), and the state required has been tucked into the `reply` function.

Some other changes:
* new `test.env`. I didn't want to add another `.env`, but I needed stuff in the test environment that isn't applicable to users. This means we can add comments to `example.env` 

closes #50 

## Did you document your work?

Some missing docs have been added to the Readme, new functionality is there too, and existing docs are updated.

## How can someone test these changes?

Steps to manually verify the change:

1. `npm i`
2. `npm t`

Link to an existing bot and make sure your existing bot works:
1. `npm link`
2. go to your bot
3. `npm link launch-vehicle-fbm`
4. run and test your bot

## What possible risks or adverse effects are there?

* The function signature for `send` is changed. But it's done in a backwards compatible change. That means that the arguments are currently introspected before being used, which adds a mess of code. We just have to hope to mature enough so the page config transition lets us delete the adapter code.
* in #49, I made `send(senderId, responseObj, pageId)` but here I changed my mind and went with `send(pageId, senderId, responseObj)` because putting `pageId` at the end looked so hacky. Now the arguments are in a more logical order.
* This seems like a lot of LOC, but it's almost all docs and tests. I'm not really happy with how many lines it is, but reading the diff, it seems reasonable.

## What are the follow-up tasks?

* ~~log a deprecation warning when bot is launched w/o page config~~
* none

## Are there any known issues?

None

## Did the test coverage decrease?

new code is covered

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -181,6 +202,7 @@ class Messenger extends EventEmitter {
       .then((session) => this.saveSession(session));
   }
 
+  // WIP
```
Comment (stripethree) [inline src/app.js]: Unclear what this WIP is referring to. Doesn't seem to match anything in the follow-up tasks...?

File: src/app.js [hunk 2]
```diff
@@ -246,14 +268,15 @@ class Messenger extends EventEmitter {
...
-    this.emit('auth', {event, senderId, session, optinRef});
+    this.emit('auth', new Response(this, {event, senderId, session, optinRef}));
     debug('onAuth for user:%d with param: %j', senderId, optinRef);
   }
 
   /*
     This is not an event triggered by Messenger, it is the post-back from the
     static Facebook login page that is made to look similar to an 'event'
   */
+  // WIP
```
Comment (stripethree) [inline src/app.js]: Ditto here.

File: src/app.js [hunk 3]
```diff
@@ -355,17 +378,25 @@ class Messenger extends EventEmitter {
     return this.cache.set(session._key, session);
   }
 
-  send(recipientId/*: string|number */, messageData/*: Object */, pageId/*: string|void */)/* Promise<Object> */ {
+  send(arg1/*: string|number */, arg2/*: string|number|Object */, arg3/*: Object|void */)/* Promise<Object> */ {
```
Comment (stripethree) [inline src/app.js]: Since this is going to be refactored into simpler logic, have you considered making the new paradigm a wrapper function for the time being and then we replace `send` down the line? This would also mean changing the third arg to the `pageAccessToken` rather than the ID, but I think that's a fair tradeoff since that is currently an optional arg and it would reduce LOC.

```
pageSend(pageId, recipientId, messageData) {
  const pageAccessToken = this.pages[pageId];
  return this.send(recipientId, messageData, pageAccessToken);
}
```

```
send(recipientId, messageData, accessToken) {
  // kills ~10 LOC here
  const pageAccessToken = accessToken || config.get('messenger.pageAccessToken');
  const options = {... }
  debug('message.send: %j', options);
  return reqPromise.post(options)
    ...
}
```
Later on, the contents of `send` can be merged into `pageSend` or `send` can be replace by `pageSend`.

Both approaches are a bit awkward TBH.  I find the above alternative easier to follow and to me, it communicates more about where the code is headed. It does come with downsides though. In either case, I think some additional comments or a link to an issue describe the next steps would provide some useful context.

*TL;DR* - I'm okay with this even though it is a bit tough to read, but wanted to propose an alternative as a thought exercise.

File: src/app.js (hunk 1 above)
Comment (crccheck) [inline src/app.js]: oh, I was marking that these methods are vestigial from the beauty lens bot and they're kinda dead code at this point but we want to bring them back eventually. I didn't want to start yak shaving to bring them up to "code" 😸 to keep this PR smaller. But I wanted to leave a mark to remind myself to stop trying to work on them.

File: src/app.js (hunk 3 above)
Comment (crccheck) [inline src/app.js]: I've been working on this concept on and off for a few weeks. I think I thought of that approach but I can't remember why I didn't stick with it. I'll give it a shot with `pageSend` 👍

**Comment by stripethree** [comment]
Tried with my test bot via `enterprise-fbm` and things are working as expected. Did not try it with the bot deployed on multiple pages.

File: src/app.js (hunk 1 above)
Comment (stripethree) [inline src/app.js]: Ok, cool. Seems like good candidates for linking to issues.

**Comment by stripethree** [comment]
Noticed that the previous version did not allow a single page to send with `reply` without a `pages` arg sent to Messenger. This is remedied with the latest updates; holding off on further review as Chris said there's a few more things coming. I'm liking what the application level bot code looks like (`enterprise-fbm`, for example below) esp. with the PR that also removes the need to call `start`.

```
const { Messenger, responses } = require('launch-vehicle-fbm');
const messenger = new Messenger();

messenger.start(); // gone soon...

messenger.on('text.greeting', ({reply, senderId, session, firstName, surName, fullName}) => {
  reply(new responses.Text(`🤖  Beep bop boop. Well hello there, ${firstName}, I am a bot.`));
});

messenger.on('text.help', ({reply}) => {
  reply(new responses.Text('🤖  I am here to offer my services to assist you.'));
});

messenger.on('text', ({reply, senderId, text, source}) => {
  reply(new responses.Text(`Echo: "${text}"`));
});

messenger.on('message.image', ({reply, url}) => {
  reply(new responses.Image(url));
});

**Comment by crccheck** [comment]
I'm moving this back to WIP to change some design. After IRL discussion, moving towards forcing `{pages}` is drastic since most bots will only have one page. It's a big step to ask people from doing `new Messenger()` to `new Messenger({pages})` so I'm backtracking that and going to keep the config based stuff now so you only need to do `new Messenger({pages})` if you need to support multiple pages.

File: test.env [hunk 4]
```diff
@@ -0,0 +1,10 @@
+FACEBOOK_APP_ID=1234567890123456
+FACEBOOK_PAGE_ID=1029384756
+MESSENGER_APP_SECRET=0123456789abcdef0123456789abcdef
+MESSENGER_VALIDATION_TOKEN=validate_me
+MESSENGER_PAGE_ACCESS_TOKEN=ThatsAReallyLongStringYouGotThere
+SERVER_URL=https://localhost/
+SLACK_WEBHOOK_URL=https://hooks.slack.com/services/T123456/B33Z/Pajamas
+SLACK_CHANNEL=channel
+LOG_FILE=/var/log/my-bot/chat.log
+ALLOW_CONFIG_MUTATIONS=Y
```
Comment (crccheck) [inline test.env]: `ALLOW_CONFIG_MUTATIONS` is the new config I needed. This will let you do `PORT=-1` in #54 without potentially confusing users.

File: src/app.js [hunk 5]
```diff
@@ -61,7 +83,14 @@ class Messenger extends EventEmitter {
...
     }
 
-    this.pages = pages;
+    if (pages && Object.keys(pages).length) {
+      this.pages = pages;
+    } else if (config.has('messenger.pageAccessToken') && config.has('facebook.pageId')) {
+      this.pages = {[config.get('facebook.pageId')]: config.get('messenger.pageAccessToken')};
+    } else {
+      this.pages = {};
+      debug("MISSING options.pages; you won't be able to reply or get profile information");
```
Comment (crccheck) [inline src/app.js]: This is my favorite line in this PR. It shows how those two config variables are actually optional depending on what your bot is doing. While the SDK doesn't exactly gracefully degrade, we could make it.

**Comment by crccheck** [comment]
I did a test w/ slack ghost just now too and 👍
